### PR TITLE
Support default option values from configuration

### PR DESCRIPTION
**Pull request by brianm**
These changes provide support for default option values from arbitrary configuration. Map and Properties based implementations of configuration providers are included.

### CODE REVIEW

File: src/main/java/org/iq80/cli/config/MapConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Map;
+
+public class MapConfiguration implements Configuration
+{
+    private final Map<String, String> config;
+
+    public <T extends Map<String, String>> MapConfiguration(T config) {
```
Comment (electrum) [inline src/main/java/org/iq80/cli/config/MapConfiguration.java]: Why is this a parameterized type?

File: src/main/java/org/iq80/cli/config/PropertiesConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    public PropertiesConfiguration(Properties properties)
+    {
+        this.properties = new Properties(properties);
+    }
+
+    public static Configuration fromProperties(File... sources) {
+        Properties current = new Properties();
+        for (File source : sources) {
+            current = new Properties(current);
```
Comment (electrum) [inline src/main/java/org/iq80/cli/config/PropertiesConfiguration.java:25]: Why is it necessary to copy the Properties object?

File: src/main/java/org/iq80/cli/config/PropertiesConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            if (source.exists()) {
+                try {
```
Comment (electrum) [inline src/main/java/org/iq80/cli/config/PropertiesConfiguration.java:27]: Extracting this into a helper method would improve readability.

File: src/main/java/org/iq80/cli/model/MetadataLoader.java [hunk 4]
```diff
@@ -167,7 +168,15 @@ public static void loadInjectionMetadata(Class<?> type, InjectionMetadata inject
                         allowedValues = null;
                     }
 
-                    OptionMetadata optionMetadata = new OptionMetadata(optionType, options, name, description, arity, required, hidden, allowedValues, path);
+                    final Optional<String> configurationKey;
+                    if (optionAnnotation.configuration() != null && optionAnnotation.configuration().length() > 0) {
```
Comment (electrum) [inline src/main/java/org/iq80/cli/model/MetadataLoader.java]: configurationKey = Optional.fromNullable(emptyToNull(optionAnnotation.configuration()));

File: src/main/java/org/iq80/cli/config/MapConfiguration.java (hunk 1 above)
Comment (brianm) [inline src/main/java/org/iq80/cli/config/MapConfiguration.java]: So you can pass ImmutableMap.of(...) without an explicit downcast to Map<String, String>

File: src/main/java/org/iq80/cli/config/PropertiesConfiguration.java (hunk 2 above)
Comment (brianm) [inline src/main/java/org/iq80/cli/config/PropertiesConfiguration.java:25]: Defensive copy --- in the ctor case we don't know who else has a reference to it.

File: src/main/java/org/iq80/cli/config/PropertiesConfiguration.java (hunk 3 above)
Comment (brianm) [inline src/main/java/org/iq80/cli/config/PropertiesConfiguration.java:27]: I disagree. It is tens lines and is not used anywhere else. Extracting it to a helper method would mostly just add a layer of indirection/obfuscation.

File: src/main/java/org/iq80/cli/model/MetadataLoader.java (hunk 4 above)
Comment (brianm) [inline src/main/java/org/iq80/cli/model/MetadataLoader.java]: Nice!

File: src/main/java/org/iq80/cli/config/MapConfiguration.java (hunk 1 above)
Comment (brianm) [inline src/main/java/org/iq80/cli/config/MapConfiguration.java]: Actually, not needed here. Using Hamcrest has messed my brain up for this stuff. Will update pull request :-)

**Comment by brianm** [comment]
Submitting new pull request based on comments from David.

**Pull request closed without merging**